articles/root-cause-analysis.md

```diff
@@ -0,0 +1,79 @@
+
+## Can We Be More Specific?
+
+Absolutely!
+
+Let's have a look at some of the painfully obvious yet all too common pitfalls.
+
+### Not Really Fixing the Issue. Or: Not Fixing The Real Issue.
+
+Do you feel some guilt reading through the three examples below? I do.
+
+* Added a fix in the wrong place. Didn't solve the problem, optionally introduced another one?
+* Fixed a `NullReferenceException` by adding another `if`, ended up with no exception but a wrong result?
+* Divided an `x` through `total` to make it relative, deployed to production, got an "unexpected" `DivideByZeroException` the day after?
+
+Things are often more subtle and not that silly. What they have in common though is that we didn't think it through all the way. We knew what _shouldn't_ happen, but didn't find out _why_ and what _should_.
+
+### "It's a Feature." Or: "We Cannot Re-Produce This!"
+
+It's a chance to improve the error message.
+
+You couldn't find customer #316 in the DB, which is expected? A `NullReferenceException` with a stack trace is still a bug. Tell them what's wrong, in their language, so everyone can understand the problem without a debugger.
+
+Also do this when you cannot re-produce a problem. Don't try to solve it based on assumptions; refine your error handling and use it to find the problem.
+
+### ... This Might Have Happened Somewhere Else
+
+"I think we might divide by `total` there, there and there too. We should check if that also throws a `DivideByZeroException`."
+
+It's great to think about it. The problem might be a pattern. But fix that separately. Equally carefully!
+
+Does this happen too often? That's what the [DRY priciple](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is about.
+
+### There's a "Real Solution" and a "Quick Solution" (and a "Workaround")
+
+After understanding the problem there might be multiple options to solve or work around it:
+
+1. adding another `if` in method x
+1. refactoring the accounting system interface to support zero totals
+1. breaking down the system into different microservices
+
+It's fine to add an `if` to get an imminent problem out of the way - after you have thought it through! In any case keep all the options in your backlog. After adding the 5th `if`, another solution could be increasingly compelling.
+
+## Conclusion
+
+All issues, even seemingly trivial ones, demand careful and thorough analysis. The problem might seem obvious, but it's always worth double-checking. You always have time for that.
+
+Having said that, be pragmatic with the fix. Don't overengineer. Decide what's needed _now_ to put out the fire, and what should be done  _later_ to prevent further blazes.
+
+But let it be an educated decision.
+
+P.S. Whatever you do. Test, test, test!
```

Comment:
Im Nachsatz kommt das Testing doch etwas schwach und spät rüber. Gibt es eine Möglichkeit, das Thema weiter oben zu verankern.
Ich liebe in diesem Zusammenhang den Satz: "Every bug is a strong indicator of a missing test."